sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1387,7 +1388,7 @@ class Analyzer(
         i.copy(right = dedupRight(left, right))
       case e @ Except(left, right, _) if !e.duplicateResolved =>
         e.copy(right = dedupRight(left, right))
-      case u @ Union(children) if !u.duplicateResolved =>
+      case u @ Union(children, _, _) if !u.duplicateResolved =>
```

Comment:
we should do this only when the by-name resolution is done?

Reply:
`duplicateResolved` checks attribute sets from children that look at `exprId` actually. By-name resolution is only for attribute name. If you think it is safer, I can add it here.

Reply:
I think it's safer to add it, to explicitly define the rule order (by-name resolution should happen before this rule)